javalib/src/main/scala/java/lang/flttodec/RyuDouble.scala

```diff
@@ -0,0 +1,446 @@
+
+      val (za, zb) = if (q <= 21) {
+        if (v % 5 == 0) (false, multipleOfPowerOf5(v, q))
+        else if (even) (multipleOfPowerOf5(u, q), false)
+        else (false, false)
+      } else {
+        (false, false)
+      }
+      (aq, bq, if (q <= 21 && !even && multipleOfPowerOf5(w, q)) cq - 1 else cq, q, za, zb)
+    } else {
+      // q = floor(-e2 * log10(5)) - 1
+      // We approximate log10(5) as a fraction: 732923 / 2^20
+      // (732923 / 1048576 = 0.69897007... which is very close)
+      val qTmp = ((-e2 * 732923) >>> 20) - 1
+      val q = if (qTmp > 0) qTmp else 0
+      val i = -e2 - q
+      val k = pow5bits(i) - Pow5BitCount
+      val j = q - k
+      val aq = mulPow5divPow2(u, i, j)
+      val bq = mulPow5divPow2(v, i, j)
+      val cq = mulPow5divPow2(w, i, j)
+
+      val (za, zb, cqAdjusted) = if (q <= 1) {
+        if (even) (isGeneralLowerBoundCase, true, cq)
+        else (false, true, cq - 1)
+      } else if (q < 63) {
+        (false, (v & ((1L << (q - 1)) - 1)) == 0, cq)
+      } else {
+        (false, false, cq)
+      }
+      (aq, bq, cq, q + e2, za, zb)
+    }
+  }
+
+  @inline
+  private def computeShortest(aq0: Long, bq0: Long, cq0: Long, za0: Boolean, zb0: Boolean,
+      e10: Int, even: Boolean, scientificNotation: Boolean): (Long, Int) = {
+    var removed = 0
+    var lastRemovedDigit = 0
+    var (aq, bq, cq) = (aq0, bq0, cq0)
+    var (za, zb) = (za0, zb0)
+
+    val output = if (za || zb) {
+      while (cq / 10 > aq / 10 &&
+          // Double.toString semantics requires printing at least two digits.
+          !(cq < 100 && scientificNotation)) {
+        za &= aq % 10 == 0
+        zb &= lastRemovedDigit == 0
+        lastRemovedDigit = (bq % 10).toInt
+        aq /= 10
+        bq /= 10
+        cq /= 10
+        removed += 1
+      }
+
+      if (za && even) {
+        while (aq % 10 == 0 &&
+            // Double.toString semantics requires printing at least two digits.
+            !(cq < 100 && scientificNotation)) {
+          zb &= lastRemovedDigit == 0
+          lastRemovedDigit = (bq % 10).toInt
+          aq /= 10
+          bq /= 10
+          cq /= 10
+          removed += 1
+        }
+      }
+      if (zb && (lastRemovedDigit == 5) && (bq % 2 == 0)) {
+        // Round even if the exact numbers is .....50..0.
+        lastRemovedDigit = 4
+      }
+      bq +  (if ((bq == aq && !(za && even)) || (lastRemovedDigit >= 5)) 1 else 0)
+    } else {
+      while (cq / 10 > aq / 10
+          // Double.toString semantics requires printing at least two digits.
+          && !(cq < 100 && scientificNotation)) {
+        lastRemovedDigit = (bq % 10).toInt
+        aq /= 10
+        bq /= 10
+        cq /= 10
+        removed += 1
+      }
+      bq + (if ((bq == aq) || (lastRemovedDigit >= 5)) 1 else 0)
+    }
+    (output, removed)
+  }
+
+  @inline
+  private def printString(output0: Long, olength: Int, exp0: Int,
+      sign: Boolean, scientificNotation: Boolean): String = {
```

Comment:
```suggestion
  private def printString(output0: Long, olength: Int, exp0: Int,
      sign: Boolean, scientificNotation: Boolean): String = {
```
This large method does not return a tuple, so I guess it's counter-productive to inline it.

Reply:
Since `printString` is called only from `doubleToString`, I guess it's not counter-productive I guess?
(The original Ryu code has inlined into the `doubleToString`, but I split into a function for readability) (embedding the printString into `doubleToString` is the best for performance maybe though)